pkg/infra/httpclient/httpclientprovider/azure_middleware.go

```diff
@@ -0,0 +1,101 @@
+
+func AzureMiddleware(cfg *setting.Cfg) httpclient.Middleware {
+	return httpclient.NamedMiddlewareFunc(azureMiddlewareName, func(opts httpclient.Options, next http.RoundTripper) http.RoundTripper {
+		if enabled, err := isAzureAuthenticationEnabled(opts.CustomOptions); err != nil {
+			return errorResponse(err)
+		} else if !enabled {
+			return next
+		}
+
+		credentials, err := getAzureCredentials(opts.CustomOptions)
+		if err != nil {
+			return errorResponse(err)
+		} else if credentials == nil {
+			credentials = getDefaultAzureCredentials(cfg)
+		}
+
+		tokenProvider, err := aztokenprovider.NewAzureAccessTokenProvider(cfg, credentials)
+		if err != nil {
+			return errorResponse(err)
+		}
+
+		scopes, err := getAzureEndpointScopes(opts.CustomOptions)
+		if err != nil {
+			return errorResponse(err)
+		}
+
+		middleware := aztokenprovider.AuthMiddleware(tokenProvider, scopes)
+		return middleware.CreateMiddleware(opts, next)
+	})
+}
+
+func errorResponse(err error) http.RoundTripper {
+	return httpclient.RoundTripperFunc(func(req *http.Request) (*http.Response, error) {
+		return nil, fmt.Errorf("invalid Azure configuration: %s", err)
+	})
+}
+
+func isAzureAuthenticationEnabled(customOptions map[string]interface{}) (bool, error) {
+	if untypedValue, ok := customOptions["azureAuth"]; !ok {
+		return false, nil
+	} else if value, ok := untypedValue.(bool); !ok {
+		err := fmt.Errorf("the field 'azureAuth' should be a bool")
+		return false, err
+	} else {
+		return value, nil
+	}
+}
+
+func getAzureCredentials(customOptions map[string]interface{}) (azcredentials.AzureCredentials, error) {
+	if untypedValue, ok := customOptions["azureCredentials"]; !ok {
+		return nil, nil
+	} else if value, ok := untypedValue.(azcredentials.AzureCredentials); !ok {
+		err := fmt.Errorf("the field 'azureCredentials' should be a valid credentials object")
+		return nil, err
+	} else {
+		return value, nil
+	}
+}
+
+func getDefaultAzureCredentials(cfg *setting.Cfg) azcredentials.AzureCredentials {
+	if cfg.Azure.ManagedIdentityEnabled {
+		return &azcredentials.AzureManagedIdentityCredentials{}
+	} else {
+		return &azcredentials.AzureClientSecretCredentials{
+			AzureCloud: cfg.Azure.Cloud,
+		}
+	}
+}
+
+func getAzureEndpointScopes(customOptions map[string]interface{}) ([]string, error) {
+	if untypedValue, ok := customOptions["azureEndpointResourceId"]; !ok {
+		err := fmt.Errorf("the field 'azureEndpointResourceId' should be set")
+		return nil, err
+	} else if value, ok := untypedValue.(string); !ok {
+		err := fmt.Errorf("the field 'azureEndpointResourceId' should be a string")
+		return nil, err
+	} else if resourceId, err := url.Parse(value); err != nil || resourceId.Scheme == "" || resourceId.Host == "" {
+		err := fmt.Errorf("invalid endpoint Resource ID URL '%s'", resourceId)
+		return nil, err
+	} else {
+		resourceId.Path = strings.TrimRight(resourceId.Path, "/") + "/.default"
+		scopes := []string{resourceId.String()}
```

Comment:
The goal of this PR is enable Azure authentication in Prometheus plugin. For now as experimental feature under a feature flag.
Once this implementation is stabilized, it will make sense for Azure Monitor datasource to start using it as well, but right now it's not the goal. I want Azure Monitor to continue using its own authentication until this is finalized to avoid regressions.

Reply:
In the experimental UI for Prometheus datasource (PR #35860) the resource ID (audience) expected to be provided by user and saved to JsonData because right now no actual endpoints or scopes are finalized yet.
This is temporary solution. In future endpoints and scopes will be predefined in the datasource (same as in Azure Monitor).

Reply:
I would not want to use this in Azure Monitor datasource for now.

Once this is stabilized and strongly-typed model introduced into `httpclient.Options`  when we can switch Azure Monitor to it as well, but for now I would want Azure Monitor to continue using its own implementation.

#37044 is one of the prerequisites for Azure Monitor. Scopes also need to be considered.